File: course_4/module_3/programming_assignment_3/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

// City represents a city with x and y coordinates.
type City struct {
	X, Y float64
}

// getDist calculates the Euclidean distance between two cities.
// It computes the square root of the sum of the squared differences
// between the x and y coordinates of the two cities.
func getDist(p1, p2 City) float64 {
	dx := p1.X - p2.X
	dy := p1.Y - p2.Y
	return math.Sqrt(dx*dx + dy*dy)
}

// loadTSPData reads the TSP data from a file and returns a slice of City objects.
// Each city has x and y coordinates parsed from the file.
// The function ensures that the number of cities declared in the file matches
// the actual number of cities provided in the file.
func loadTSPData(filePath string) ([]City, error) {
	// Open the file for reading.
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	// Read the first line, which indicates the number of cities.
	scanner.Scan()
	nCities, err := strconv.Atoi(strings.TrimSpace(scanner.Text()))
	if err != nil {
		return nil, fmt.Errorf("invalid number of cities: %v", err)
	}

	// Create a slice to hold the cities.
	cities := make([]City, 0, nCities)

	// Read the subsequent lines, each containing city coordinates.
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		// Split the line into fields.
		parts := strings.Fields(line)
		if len(parts) != 3 { // Ensure each line has three parts: index, x, y.
			return nil, fmt.Errorf("invalid city data: %s", line)
		}

		// Parse the x and y coordinates.
		x, errX := strconv.ParseFloat(parts[1], 64)
		y, errY := strconv.ParseFloat(parts[2], 64)
		if errX != nil || errY != nil {
			return nil, fmt.Errorf("invalid coordinates: %s", line)
		}

		// Append the city to the slice.
		cities = append(cities, City{X: x, Y: y})
	}

	// Check that the number of cities matches the declared count.
	if len(cities) != nCities {
		return nil, fmt.Errorf("mismatch between declared and actual number of cities")
	}

	return cities, nil
}

// nearestNeighborTSP solves the Traveling Salesman Problem using the nearest neighbor heuristic.
// The algorithm starts at the first city and repeatedly visits the closest unvisited city.
// Once all cities are visited, it returns to the starting city to complete the tour.
// The function returns the total distance of the computed tour.
func nearestNeighborTSP(cities []City) float64 {
	nCities := len(cities)           // Total number of cities.
	visited := make([]bool, nCities) // Track whether each city has been visited.
	currCity := 0                    // Start at the first city (index 0).
	visited[currCity] = true         // Mark the starting city as visited.
	nVisited := 1                    // Number of cities visited so far.
	cost := 0.0                      // Total distance of the tour.

	// While there are still unvisited cities, find the nearest neighbor.
	for nVisited < nCities {
		nextCity := -1             // Index of the next city to visit.
		minDist := math.MaxFloat64 // Minimum distance to the next city.

		// Iterate through all cities to find the closest unvisited city.
		for i := 0; i < nCities; i++ {
			if visited[i] { // Skip cities that have already been visited.
				continue
			}
			// Calculate the distance from the current city to city i.
			dist := getDist(cities[currCity], cities[i])
			// Update the next city if this city is closer.
			if dist < minDist {
				minDist = dist
				nextCity = i
			}
		}

		// Add the distance to the total cost.
		cost += minDist
		// Mark the next city as visited.
		visited[nextCity] = true
		// Move to the next city.
		currCity = nextCity
		// Increment the count of visited cities.
		nVisited++
	}

	// Add the distance to return to the starting city to complete the tour.
	cost += getDist(cities[currCity], cities[0])
	return cost
}

func main() {
	// Define the path to the TSP data file.
	filePath := "course_4/module_3/programming_assignment_3/nn.txt"

	// Load the TSP data from the file.
	cities, err := loadTSPData(filePath)
	if err != nil {
		// Print an error message and exit if there is an issue with loading data.
		fmt.Printf("Error loading TSP data: %v\n", err)
		return
	}

	// Solve the TSP using the nearest neighbor heuristic.
	totalCost := nearestNeighborTSP(cities)

	// Output the total distance of the tour, rounded down to the nearest integer.
	fmt.Printf("Total Distance: %d\n", int(math.Floor(totalCost)))
}

File: course_4/module_3/programming_assignment_3/solution.py
import math


class City:
    """
    Represents a city with x and y coordinates.
    """
    def __init__(self, x: float, y: float):
        self.x = x  # X-coordinate of the city.
        self.y = y  # Y-coordinate of the city.

    def distance_to(self, other: "City") -> float:
        """
        Calculates the Euclidean distance to another city.
        """
        dx = self.x - other.x
        dy = self.y - other.y
        return math.sqrt(dx * dx + dy * dy)


def load_tsp_data(file_path: str) -> list[City]:
    """
    Loads the TSP data from the file and returns a list of City objects.
    """
    with open(file_path, "r") as file:
        lines = file.readlines()

    # Read the number of cities from the first line.
    n_cities = int(lines[0].strip())

    # Parse city coordinates.
    cities = []
    for line in lines[1:]:
        parts = line.strip().split()
        if len(parts) == 3:
            _, x, y = map(float, parts)  # Skip the city index.
            cities.append(City(x, y))

    # Validate the number of cities matches the declared count.
    if len(cities) != n_cities:
        raise ValueError("Mismatch between declared and actual number of cities.")

    return cities


def nearest_neighbor_tsp(cities: list[City]) -> float:
    """
    Solves the Traveling Salesman Problem using the nearest neighbor heuristic.
    Returns the total distance of the computed tour.
    """
    n_cities = len(cities)
    visited = [False] * n_cities  # Track whether each city has been visited.
    curr_city_index = 0  # Start at the first city.
    visited[curr_city_index] = True
    n_visited = 1
    total_distance = 0.0

    while n_visited < n_cities:
        next_city_index = -1
        min_dist = float("inf")

        # Find the nearest unvisited city.
        for i in range(n_cities):
            if not visited[i]:
                dist = cities[curr_city_index].distance_to(cities[i])
                if dist < min_dist:
                    min_dist = dist
                    next_city_index = i

        # Move to the next city.
        total_distance += min_dist
        visited[next_city_index] = True
        curr_city_index = next_city_index
        n_visited += 1

    # Add the distance to return to the starting city.
    total_distance += cities[curr_city_index].distance_to(cities[0])
    return total_distance


def main():
    file_path = "nn.txt"

    try:
        # Load the TSP data.
        cities = load_tsp_data(file_path)

        # Solve the TSP using the nearest neighbor heuristic.
        total_distance = nearest_neighbor_tsp(cities)

        # Output the total distance, rounded down to the nearest integer.
        print(f"Total Distance: {int(math.floor(total_distance))}")
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()
